### PR TITLE
Add solutions to chapter 4

### PR DESCRIPTION
**Pull request by alternateved**
### Solutions for Chapter 4

Thank you for that course. I have to admit, it was pretty demanding. 
cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -471,11 +485,12 @@ Applicatives can be found in many applications:
...
-    pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+  pure :: a -> Secret e a
+  pure = Reward
 
-    (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+  (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
+  (<*>) (Trap y) _ = Trap y
+  (<*>) (Reward f) x = fmap f x
```
Comment (vrom911) [inline src/Chapter4.hs:493]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -488,7 +503,17 @@ Implement the 'Applicative' instance for our 'List' type.
...
+instance Applicative List where
+  pure :: a -> List a
+  pure x = Cons x Empty
+
+  (<*>) :: List (a -> b) -> List a -> List b
+  (<*>) Empty _ = Empty
+  (<*>) (Cons f fs) xs = concatList (fmap f xs) ((<*>) fs xs)
 
+concatList :: List a -> List a -> List a
+concatList Empty ys = ys
```
Comment (vrom911) [inline src/Chapter4.hs:515]: Perfect!

You can add one more case to make your solution even more efficient:

```haskell
append xs Empty = xs
```

In that case, you won't need to walk through each element of the first list when the second one is empty and return the list instantly 🚤

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask @chshersh or me any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

<hr>

If you would like to continue your journey to Haskell, check out our [Haskell Knowledge Map](https://kowainik.github.io/images/Haskell_Knowledge_Map.png) for the list of topics to explore 🗺️

File: src/Chapter4.hs (hunk 1 above)
Comment (alternateved) [reply src/Chapter4.hs:493]: Thanks!

File: src/Chapter4.hs (hunk 2 above)
Comment (alternateved) [reply src/Chapter4.hs:515]: Oh, I see. I'll do it then.
```suggestion
concatList Empty ys = ys
concatList ys Empty = ys
```

**Pull request merged**